drivers: i2c_dw: add support for agilex platform and reset

samitsingh7 opened the pull request:
- Updated driver to reset the i2c node if its corresponding dtsi instance has a resets property.
- Handling invalid 0 length i2c transfer and logging error if i2c transfer is initiated with length 0.
- Modified i2c shell to support i2c scan command only if I2C_ZERO_LEN_XFER_SUPPORTED Kconfig flag is enabled.
- Updated i2c shell for checking i2c-speed parameter in i2c speed shell command.

samitsingh7 committed:
dts: arm64: intel: intel_socfpga_agilex: Adding device tree for i2c

Added device tree for I2C

Signed-off-by: samit singh sikarwar <samit.singh.sikarwar@intel.com>

samitsingh7 committed:
drivers: i2c: i2c_dw: adding optional feature support and improvements

Handling invalid 0 length i2c transfer and logging error
if i2c transfer is initiated with length 0.
Modified i2c shell to support i2c scan command only if
I2C_ZERO_LEN_XFER_SUPPORTED Kconfig flag is enabled.
Updated i2c shell for checking i2c-speed parameter in
i2c speed shell command.

Signed-off-by: samit singh sikarwar <samit.singh.sikarwar@intel.com>

samitsingh7 commented on drivers/i2c/i2c_dw.c:
```diff
@@ -709,6 +753,8 @@ static int i2c_dw_set_master_mode(const struct device *dev)
 	write_tx_tl(ic_comp_param_1.bits.tx_buffer_depth + 1, reg_base);
 	write_rx_tl(ic_comp_param_1.bits.rx_buffer_depth + 1, reg_base);
 
+	LOG_DBG("I2C: Host registered as Master Device");
```
Updated, thanks!

samitsingh7 commented on drivers/i2c/i2c_dw.c:
```diff
@@ -736,7 +782,7 @@ static int i2c_dw_set_slave_mode(const struct device *dev, uint8_t addr)
 	write_rx_tl(0, reg_base);
 
-	LOG_DBG("I2C: Host registed as Slave Device");
+	LOG_DBG("I2C: Host registered as Slave Device");
```
Updated, thanks!

samitsingh7 commented on drivers/i2c/i2c_shell.c:
```diff
@@ -289,6 +291,12 @@ static int cmd_i2c_speed(const struct shell *shell_ctx, size_t argc, char **argv
 
 	speed = strtol(argv[ARGV_DEV + 1], NULL, 10);
+
+	if (speed > I2C_SPEED_DT) {
```
i2c shell command: i2c speed `<device>` `<input>`
will result in setting up the same speed either if we pass input as (1 or 9), (2 or 10), (3 or 11)..
Because in i2c speed shell function, I2C_SPEED_SET(speed) macro masks the speed parameter with 0x7.

Thus, introduced the check for the input parameter as passing the value more than 0x7 is not possible.

samitsingh7 commented on drivers/i2c/i2c_shell.c:
```diff
@@ -33,6 +33,7 @@ static int get_bytes_count_for_hex(char *arg)
 	return MIN(MAX_BYTES_FOR_REGISTER_INDEX, length);
 }
 
+#if CONFIG_I2C_SMBUS_SUPPORTED
```
i2c scan shell command initiates i2c transfer without any data which is not a standard in i2c protocol but SMbus supports such transfers.
Also, the current i2c_dw driver is not handling 0 length i2c transfer. Thus, issuing this shell command could hang the system(which is happening in i2c_dw driver).

So by introducing kconfig macro: CONFIG_I2C_SMBUS_SUPPORTED,
By default, this macro is enabled and the i2c scan shell command can be issued but if we disable this macro explicitly then the i2c scan shell command will not be available and thus preventing the program from getting hung if such i2c transfers without data are not handled in the driver.

teburd replied:
So if I'm understanding this correctly, the issue is on this device the zero sized transfer used by i2c scan here https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_shell.c#L79 is unsupported by this particular hardware or driver.

There's nothing in the i2c specification regarding sending an address only being invalid that I can find. Do you have a source for this being invalid?

The Kconfig here makes sense in terms of this particular hardware and driver combo possibly not supporting i2c scan. But to have a Kconfig like this is a bit strange to me. Instead the driver should return an error code on such transfers and this Kconfig is no longer needed I'd think.

If it's something the driver can possibly do on some instances of the designware controller then it should be a Kconfig along the lines of I2C_DW_HAS_ZERO_LEN_XFER or so and worked with in that way.

samitsingh7 replied:
> There's nothing in the i2c specification regarding sending an address only being invalid that I can find. Do you have a source for this being invalid?

I agree that sending an address only in i2c transaction (Quick Command) is valid.

As per the comment of the i2c-scan function:-
```
WARNING: As there is no standard I2C detection command, this code
 uses arbitrary SMBus commands (namely SMBus quick write and SMBus
  receive byte) to probe for devices.  This operation can confuse
 your I2C bus, cause data loss, and is known to corrupt the Atmel
  AT24RF08 EEPROM found on many IBM Thinkpad laptops. 
  https://manpages.debian.org/buster/i2c-tools/i2cdetect.8.en.html
```
 We are issuing SMbus Quick command for i2c-scan functionality. Also, it is well documented that this operation could confuse i2c bus and cause data loss. or corrupt the target device. The same is documented here: https://manpages.debian.org/buster/i2c-tools/i2cdetect.8.en.html#WARNING. 

Another such combination where this was not working: https://github.com/zephyrproject-rtos/zephyr/pull/54694.

Also, few of the I2C IPs support Quick Commands or 0 length i2c command based on the hardware configuration.
Thus, making use of this Kconfig macro we can avoid such conditions from various i2c controller and target device combination on various platform by disabling i2c-scan functionality for such applications where 0 length command is not supported.


> The Kconfig here makes sense in terms of this particular hardware and driver combo possibly not supporting i2c scan. 

Exactly as mentioned above and observed from various OS and hardware combination that i2c scan command could cause an issue. Thus introducing Kconfig macro which could help in enabling and disabling for the combination user have.


>Instead the driver should return an error code on such transfers and this Kconfig is no longer needed I'd think.

Totally agreed and doing the same here, we are doing 0 length check and  returning an error code - https://github.com/samitsingh7/zephyr_main/blob/i2c_dw_driver/drivers/i2c/i2c_dw.c#L478
The introduction of this Kconfig is rather for facilitating user to disable the i2c scan functionality from shell and safeguard from vulnerabilities from data corruption or i2c bus hang.

There could be different platforms where the i2c controller is not configured to support i2c quick commands.
Then all such platforms can make use of this macro to enable or disable this shell command.
We are still open to suggestions on alternative names for this macro: I2C_ZERO_LEN_XFER_SUPPORTED or I2C_QUICK_CMD_EN ?

Thanks for reviewing the changes!

samitsingh7 replied:
Requested changes done

samitsingh7 commented on drivers/i2c/i2c_shell.c:
```diff
@@ -320,10 +328,12 @@ static void device_name_get(size_t idx, struct shell_static_entry *entry)
 SHELL_DYNAMIC_CMD_CREATE(dsub_device_name, device_name_get);
 
 SHELL_STATIC_SUBCMD_SET_CREATE(sub_i2c_cmds,
+#if CONFIG_I2C_SMBUS_SUPPORTED
```
Macro name: **_CONFIG_I2C_SMBUS_SUPPORTED_**, is chosen like this because i2c transfer without any data is supported by SMbus.

By default, this macro is enabled and the i2c scan shell command will be available for an execution but if we disable this macro explicitly then the i2c scan shell command will not be available and thus preventing the program from getting hung if  i2c transfers without data are not handled in the driver (ex: i2c_dw driver).

teburd commented:
Renaming of the Kconfig option is a +1 from me

samitsingh7 commented:
I would like to have the stale label removed

samitsingh7 commented:
> Renaming of the Kconfig option is a +1 from me

Done

samitsingh7 commented:
I would like to have the stale label removed and requesting to add **dev-review** label.

teburd commented:
This will need a rebase to fix the conflicts, I'll try and get a few reviewers looking once that's fixed

henrikbrixandersen commented on drivers/i2c/i2c_dw.c:
```diff
@@ -635,6 +637,21 @@ static int i2c_dw_transfer(const struct device *dev,
 		return ret;
 	}
 
+	/*
+	 * I2C transfer with 0 length are invalid but
+	 * SMbus quick commands uses 0 length transfer.
+	 * And in order to support SMbus quick commands, the driver
+	 * has to be enhanced.
+	 */
```
This sounds like a feature appropriate for `drivers/smbus/`, not `drivers/i2c/`? Same goes for the Kconfig introduced?

samitsingh7 replied:
Few of the I2C IPs supports Quick Commands or 0 length i2c commands based on the hardware configuration.

henrikbrixandersen commented on drivers/i2c/i2c_dw.c:
```diff
@@ -878,6 +901,8 @@ static int i2c_dw_set_master_mode(const struct device *dev)
 	write_tx_tl(ic_comp_param_1.bits.tx_buffer_depth + 1, reg_base);
 	write_rx_tl(ic_comp_param_1.bits.rx_buffer_depth + 1, reg_base);
 
+	LOG_DBG("I2C: Host registered as Controller Device");
+
```
No need for this.

samitsingh7 replied:
Removed.

henrikbrixandersen commented on drivers/i2c/i2c_dw.h:
```diff
@@ -32,6 +32,7 @@ extern "C" {
 typedef void (*i2c_isr_cb_t)(const struct device *port);
 
 
+
```
Stray newline.

samitsingh7 replied:
Now, not there

henrikbrixandersen commented on drivers/i2c/Kconfig:
```diff
@@ -23,6 +23,12 @@ config I2C_SHELL
 	  The I2C shell supports scanning, bus recovery, I2C read and write
 	  operations.
 
+config I2C_ZERO_LEN_XFER_SUPPORTED
+	bool "I2C transfer without data (data length 0)"
+	default y
+	help
+	  Enable this flag to support 0 length I2C transfers(quick commands).
```
Quick commands are SMBus, not I2C.

samitsingh7 replied:
As, few of the I2C IPs support Quick Commands based on the hardware configuration (for various i2c-controller and target device combination on various platforms).

Thus, making use of this Kconfig macro we can enable/disable the support of functionalities or commands which are making use of 0 data-length i2c commands.

henrikbrixandersen reviewed (changes requested):
I think this PR should be split in multiple PRs:
- One for the new platform support
- One for the Quick Command feature (which, to be honest, I do not think belongs in an I2C driver. This should be a feature of an SMBus driver).

samitsingh7 commented:
> I think this PR should be split in multiple PRs:
> 
> * One for the new platform support
> * One for the Quick Command feature (which, to be honest, I do not think belongs in an I2C driver. This should be a feature of an SMBus driver).

The main objective of this PR is to add platform support. Additionally, we've included checks to enable/disable functionalities or commands that utilize 0 data-length I2C commands. This serves as a crucial preventative measure, particularly because certain I2C IPs support Quick Commands based on hardware configuration. 

It's important to note that no support for Quick commands feature has been added; rather, we've introduced Kconfig Macro checks only.

Your thorough review of these changes is greatly appreciated.

samitsingh7 commented on drivers/i2c/i2c_dw.c:
```diff
@@ -458,10 +458,12 @@ static void i2c_dw_isr(const struct device *port)
+			    (i2c_dw_read_byte_non_blocking(port, &data) == 0)) {
 				slave_cb->write_received(dw->slave_cfg, data);
 			}
+
```
done

henrikbrixandersen replied:
No?

samitsingh7 replied:
Done

samitsingh7 commented:
Hi @henrikbrixandersen ,

Requesting your review on this PR.

henrikbrixandersen reviewed (changes requested):
> Hi @henrikbrixandersen ,
> 
> Requesting your review on this PR.

I already reviewed it, but you have not addressed my comments? I still do not support adding SMBus-specific features to our I2C drivers. We have an SMBus driver class for this.

samitsingh7 commented:
I would like to have the stale label removed
@teburd

samitsingh7 commented:
I would like to have the stale label removed
@teburd

samitsingh7 commented:
I would like to have the stale label removed
@teburd

teburd commented:
> I would like to have the stale label removed @teburd

it was already done, are you planning on addressing the comments from @henrikbrixandersen otherwise it will continue to go stale, I'd like to see some progress here before the bot marks it stale again. Removing the stale tag won't move it forward.

samitsingh7 commented:
Hi @henrikbrixandersen ,

Requesting your review of this PR to take it forward.
I have addressed all your comments.

samitsingh7 commented on CODEOWNERS:
```diff
@@ -496,7 +496,7 @@
 /drivers/i2c/i2c_rv32m1_lpi2c*            @henrikbrixandersen
 /drivers/i2c/*sam0*                       @Sizurka
-/drivers/i2c/i2c_dw*                      @dcpleung
+/drivers/i2c/i2c_dw*                      @dcpleung @samitsingh7
```
I’ve retained this commit since it poses no issues in its current state.

samitsingh7 commented:
> > I would like to have the stale label removed @teburd
> 
> it was already done, are you planning on addressing the comments from @henrikbrixandersen otherwise it will continue to go stale, I'd like to see some progress here before the bot marks it stale again. Removing the stale tag won't move it forward.

I have addressed the comments from @henrikbrixandersen.

henrikbrixandersen commented on drivers/i2c/Kconfig:
```diff
@@ -22,6 +22,12 @@ config I2C_SHELL
 
+config I2C_ZERO_LEN_XFER_SUPPORTED
+	bool "I2C transfer without data (data length 0)"
+	default y
```
Shouldn't this be disabled by default and selected by drivers supporting zero-length transfers?

samitsingh7 replied:
Some specific hardware configuration does not support zero-length transfers. In such cases,  explicitly disable it. However, the default behavior should remain enabled to avoid impacting other configurations unnecessarily, given the current default is set to 'enabled.'"

henrikbrixandersen commented on drivers/i2c/i2c_dw.c:
```diff
@@ -738,16 +753,15 @@ static int i2c_dw_transfer(const struct device *dev, struct i2c_msg *msgs, uint8
 
 static int i2c_dw_runtime_configure(const struct device *dev, uint32_t config)
 {
-	struct i2c_dw_dev_config *const dw = dev->data;
-	uint32_t value = 0U;
-	uint32_t rc = 0U;
+	struct i2c_dw_dev_config * const dw = dev->data;
+	uint32_t	value = 0U;
+	int		ret = 0;
```
We usually do not indent variable names like this in Zephyr. Please leave this change out of this PR.

samitsingh7 replied:
done

henrikbrixandersen commented on drivers/i2c/i2c_dw.c:
```diff
@@ -904,7 +925,6 @@ static int i2c_dw_set_master_mode(const struct device *dev)
 
 	write_tx_tl(ic_comp_param_1.bits.tx_buffer_depth + 1, reg_base);
 	write_rx_tl(ic_comp_param_1.bits.rx_buffer_depth + 1, reg_base);
-
```
Please leave the empty line before the return statement intact.

henrikbrixandersen commented on drivers/i2c/i2c_dw.c:
```diff
@@ -1208,19 +1226,25 @@ static int i2c_dw_initialize(const struct device *dev)
 	(.dma_dev = DEVICE_DT_GET(DT_INST_DMAS_CTLR_BY_IDX(n, 0)),),	\
 	())), ())
 
-#define I2C_DEVICE_INIT_DW(n)                                                                      \
-	PINCTRL_DW_DEFINE(n);                                                                      \
-	I2C_PCIE_DEFINE(n);                                                                        \
-	static void i2c_config_##n(const struct device *port);                                     \
-	static const struct i2c_dw_rom_config i2c_config_dw_##n = {                                \
-		I2C_CONFIG_REG_INIT(n).config_func = i2c_config_##n,                               \
-		.bitrate = DT_INST_PROP(n, clock_frequency),                                       \
-		RESET_DW_CONFIG(n) PINCTRL_DW_CONFIG(n) I2C_DW_INIT_PCIE(n)                        \
-			I2C_CONFIG_DMA_INIT(n)};                                                   \
-	static struct i2c_dw_dev_config i2c_##n##_runtime;                                         \
-	I2C_DEVICE_DT_INST_DEFINE(n, i2c_dw_initialize, NULL, &i2c_##n##_runtime,                  \
-				  &i2c_config_dw_##n, POST_KERNEL, CONFIG_I2C_INIT_PRIORITY,       \
-				  &funcs);                                                         \
-	I2C_DW_IRQ_CONFIG(n)
+#define I2C_DEVICE_INIT_DW(n)                                                 \
+	PINCTRL_DW_DEFINE(n);                                                 \
+	I2C_PCIE_DEFINE(n);                                                   \
+	static void i2c_config_##n(const struct device *port);                \
+	static const struct i2c_dw_rom_config i2c_config_dw_##n = {           \
+		I2C_CONFIG_REG_INIT(n)                                        \
+		.config_func = i2c_config_##n,                                \
+		.bitrate = DT_INST_PROP(n, clock_frequency),                  \
+		RESET_DW_CONFIG(n)                                            \
+		PINCTRL_DW_CONFIG(n)                                          \
+		I2C_DW_INIT_PCIE(n)                                           \
+		I2C_CONFIG_DMA_INIT(n)					      \
+	};                                                                    \
+	static struct i2c_dw_dev_config i2c_##n##_runtime;                    \
+	I2C_DEVICE_DT_INST_DEFINE(n, i2c_dw_initialize, NULL,                 \
+			      &i2c_##n##_runtime, &i2c_config_dw_##n,         \
+			      POST_KERNEL, CONFIG_I2C_INIT_PRIORITY,          \
+			      &funcs);                                        \
+	I2C_DW_IRQ_CONFIG(n)                                          \
+	\
```
Please avoid mixing reformatting with functional changes.

henrikbrixandersen commented on drivers/i2c/i2c_dw.c:
```diff
@@ -839,14 +853,17 @@ static int i2c_dw_runtime_configure(const struct device *dev, uint32_t config)
 
 			dw->hcnt = value;
 		} else {
-			rc = -EINVAL;
+			ret = -EINVAL;
```
This change looks unrelated to the actual, functional change. Please leave this kind of changes out as it makes reading the diff much harder.

samitsingh7 replied:
Yes, the change is unrelated to functionality.
The variable name 'rc' was not very intuitive, so I updated it to 'ret' for clarity.
In the future, I’ll avoid making such changes.

henrikbrixandersen commented:
> Hi @henrikbrixandersen-
> 
> Addressed all the comments.
> 
> Seeking your review.

No, you did not address all of my comments. You are still adding support for Quick Commands, which is an SMBus feature, to the I2C API, which I commented on earlier, while we have a dedicated SMBus API.

samitsingh7 commented:
> You are still adding support for Quick Commands, which is an SMBus feature, to the I2C API, which I commented on earlier, while we have a dedicated SMBus API.

I understand your concern regarding SMBus Quick Commands, but I am not adding support for SMBus features to the I2C API. The purpose of the I2C_ZERO_LEN_XFER_SUPPORTED macro is solely to provide a way to prevent system hangs in drivers that do not handle zero-length I2C transfers (such as i2c_dw).

The existing i2c scan shell command initiates a transfer with zero-length data, which is not standard in I2C but is supported in SMBus. Since the i2c_dw driver does not handle this case properly, issuing the command can lead to a system hang. The macro allows users to disable the command explicitly, avoiding this issue without modifying I2C behavior.

This change does not introduce SMBus support into the I2C API but rather provides a safeguard against unintended hangs in non-SMBus-capable I2C drivers. Let me know if further clarification is needed.

henrikbrixandersen commented on drivers/i2c/i2c_shell.c:
```diff
@@ -33,6 +33,7 @@ static int get_bytes_count_for_hex(char *arg)
 	return MIN(MAX_BYTES_FOR_REGISTER_INDEX, length);
 }
 
+#if CONFIG_I2C_ZERO_LEN_XFER_SUPPORTED
```
I still do not understand why the Kconfig is needed. The driver returns an error on unsupported msg lengths. Why isn't this sufficient?

samitsingh7 replied:
Hi @henrikbrixandersen ,

I have explained the reasoning behind that here: https://github.com/zephyrproject-rtos/zephyr/pull/60427#discussion_r1265577497

samitsingh7 replied:
Hi @henrikbrixandersen, Please let me know if anything else is blocking this PR.

soburi commented on drivers/i2c/Kconfig:
```diff
@@ -22,6 +22,12 @@ config I2C_SHELL
 	  The I2C shell supports scanning, bus recovery, I2C read and write
 	  operations.
 
+config I2C_ZERO_LEN_XFER_SUPPORTED
```
If this config use only in i2c_sell.c, the name should be better start with I2C_SHELL_.
